src/controller/java/AndroidCommissioningWindowOpener.cpp

```diff
@@ -0,0 +1,176 @@
+/*
+ *    Copyright (c) 2022 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "AndroidCommissioningWindowOpener.h"
+
+#include <app-common/zap-generated/cluster-objects.h>
+#include <controller-clusters/zap-generated/CHIPClusters.h>
+#include <lib/core/CHIPSafeCasts.h>
+#include <lib/support/CHIPMem.h>
+#include <lib/support/JniReferences.h>
+#include <lib/support/JniTypeWrappers.h>
+#include <protocols/secure_channel/PASESession.h>
+#include <setup_payload/ManualSetupPayloadGenerator.h>
+#include <setup_payload/QRCodeSetupPayloadGenerator.h>
+
+using namespace chip::app::Clusters;
+using namespace chip::System::Clock;
+
+namespace {
+jmethodID sOnSuccessMethod = nullptr;
+jmethodID sOnErrorMethod   = nullptr;
+} // namespace
+namespace chip {
+namespace Controller {
+
+AndroidCommissioningWindowOpener::AndroidCommissioningWindowOpener(DeviceController * controller, jobject jCallbackObject) :
+    CommissioningWindowOpener(controller), mOnOpenCommissioningWindowCallback(OnOpenCommissioningWindowResponse, this),
+    mOnOpenBasicCommissioningWindowCallback(OnOpenBasicCommissioningWindowResponse, this)
+{
+    JNIEnv * env  = JniReferences::GetInstance().GetEnvForCurrentThread();
+    mJavaCallback = env->NewGlobalRef(jCallbackObject);
+
+    jclass callbackClass = env->GetObjectClass(jCallbackObject);
+
+    sOnSuccessMethod = env->GetMethodID(callbackClass, "onSuccess", "(JLjava/lang/String;Ljava/lang/String;)V");
```

Comment:
since this is setting global variables, what guarantees that AndroidCommissioningWindowOpener is indeed a singleton and is only instantiated once?

why are these not member variables instead?

Reply:
I understand your comment. I will fix these to member variables.